nixos/doc/manual/configuration/matrix.xml

```diff
@@ -0,0 +1,183 @@
+   </link>.
+   <programlisting>
+    let
+      fqdn = let
+        join = hostName: domain: hostName + optionalString (domain != null) ".${domain}";
+        in join config.networking.hostName config.networking.domain;
+
+    in {
+      networking = {
+        hostName = "jellyfish";
+        domain = "nixos.org";
+      };
+      networking.firewall.allowedTCPPorts = [ 80 443 ];
+
+      services.nginx = {
+        enable = true;
+        # only recommendedProxySettings is strictly required, but the rest makes sense as well
+        recommendedTlsSettings = true;
+        recommendedGzipSettings = true;
+        recommendedOptimisation = true;
+        recommendedProxySettings = true;
+
+        virtualHosts = {
+          # This host section can be placed on a different host than the rest,
+          # i.e. to delegate from the host being accessible as ${config.networking.domain}
+          # to another host actually running the Matrix homeserver.
+          "${config.networking.domain}" = {
+            locations."= /.well-known/matrix/server".extraConfig = let
+              # use 443 instead of the default 8448 port to unite
+              # the client-server and server-server port for simplicity
+              server = { "m.server" = "${fqdn}:443"; };
+            in ''
+              add_header Content-Type application/json;
+              return 200 '${builtins.toJSON server}';
+            '';
+            locations."= /.well-known/matrix/client".extraConfig = let
+              client = {
+                "m.homeserver" =  { "base_url" = "https://${fqdn}"; };
+                "m.identity_server" =  { "base_url" = "https://vector.im"; };
+              };
+            # ACAO required to allow riot-web on any URL to request this json file
+            in ''
+              add_header Content-Type application/json;
+              add_header Access-Control-Allow-Origin *;
+              return 200 '${builtins.toJSON client}';
+            '';
+          };
+
+          # Reverse proxy for Matrix client-server and server-server communication
+          ${fqdn} = {
+            enableACME = true;
+            addSSL = true;
+            forceSSL = true;
+
+            # or do a redirect instead of the 404, or whatever is appropriate for you
+            locations."/".extraConfig = ''
+              return 404;
+            '';
+
+            # forward all Matrix API calls to the synapse Matrix homeserver
+            locations."/_matrix" = {
+              proxyPass = "http://[::1]:8008";
+            };
+          };
+        };
+      };
+      services.matrix-synapse = {
+        enable = true;
+        server_name = config.networking.domain;
+        listeners = [
+          {
+            port = 8008;
+            bind_address = "::1";
+            type = "http";
+            tls = false;
+            x_forwarded = true;
+            resources = [
+              { names = ["client"]; compress = true; }
```

Comment:
Isn't compression already done by the reverse-proxy for clients too?

Reply:
With `recommendedGzipSettings`, that is indeed the case, thanks. :+1: Did not put enough research in it.

Combined the lines to a single one and noted that `recommendedGzipSettings` is required as well.